gobblin-runtime/src/main/java/gobblin/runtime/GobblinMultiTaskAttempt.java

```diff
@@ -0,0 +1,127 @@
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
+
+import gobblin.annotation.Alpha;
+import gobblin.commit.CommitStep;
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.metastore.StateStore;
+import gobblin.source.workunit.WorkUnit;
+import gobblin.util.ExecutorsUtils;
+import gobblin.util.executors.IteratorExecutor;
+
+import javax.annotation.Nullable;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * Attempt of running multiple {@link Task}s generated from a list of{@link WorkUnit}s.
+ * A {@link GobblinMultiTaskAttempt} is usually a unit of workunits that are assigned to one container.
+ */
+@Slf4j
+@RequiredArgsConstructor
+@Alpha
+public class GobblinMultiTaskAttempt {
+  private final List<WorkUnit> workUnits;
+  private final String jobId;
+  private final JobState jobState;
+  private final TaskStateTracker taskStateTracker;
+  private final TaskExecutor taskExecutor;
+  private final Optional<String> containerIdOptional;
+  private final Optional<StateStore<TaskState>> taskStateStoreOptional;
+  private List<Task> tasks;
+
+  // Additional commit steps that may be added by different launcher, and can be environment specific.
+  private List<CommitStep> additionalCommitSteps;
+
+  /**
+   * Run {@link #workUnits} assigned in this attempt.
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  public void run()
+      throws IOException, InterruptedException {
+    // TODO -  add workunits execution.
+  }
+
+  /**
+   * Commit {@link #tasks} by 1. calling {@link Task#commit()} in parallel; 2. executing any additional {@link CommitStep};
+   * 3. persist task statestore.
+   * @throws IOException
+   */
+  public void commit()
+      throws IOException {
+    Iterator<Callable<Void>> callableIterator =
+        Iterators.transform(this.tasks.iterator(), new Function<Task, Callable<Void>>() {
+          @Override
+          public Callable<Void> apply(final Task task) {
+            return new Callable<Void>() {
+              @Nullable
+              @Override
+              public Void call()
+                  throws Exception {
+                task.commit();
+                return null;
+              }
+            };
+          }
+        });
+
+    try {
+      List<Future<Void>> futures = new IteratorExecutor<>(callableIterator, this.getTaskCommitThreadPoolSize(),
+          ExecutorsUtils.newDaemonThreadFactory(Optional.of(log), Optional.of("Task-committing-pool-%d"))).execute();
+
+      for (Future<Void> future : futures) {
+        try {
+          future.get();
+        } catch (ExecutionException exc) {
+          log.error("Failed to commit a task.", exc.getCause());
+        }
+      }
+    } catch (InterruptedException ie) {
+      log.error("Committing of tasks interrupted. Aborting.");
+      throw new RuntimeException(ie);
+    }
+
+    for (CommitStep additionalCommitStep: this.additionalCommitSteps) {
```

Comment:
It is actually the "clean up" steps, which should be always executed. I will put it in a finally cause.